opt: push down null rejection filters to inputs in join expression

yongyanglai opened the pull request:
opt: push down null rejection filters to inputs in join expression

I add a new rewriting rule [PushDownNullRejectionForJoin]. The whole work contains three parts:
```
1.  Add support for operators <Scan> and <Select> to derive null rejection columns
2.  Add rule [PushDownNullRejectionForJoin] and corresponding logical test
3.  Update other logical tests
```
This pull request is split into two commits, the first commit include first two parts of work, the second contains the third part of work.

Release note: None
Fixes #20100

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50186)
<!-- Reviewable:end -->

yongyanglai commented:
I made rule [PushDownNullRejectionForJoin] a exploration rule because I found not all queries benefit from pushing down null rejection. Make it a normalization rule will push down every possible null rejection filters but if columns contains few null values, it is not worth doing so.(I found some existing push-down rules have the same problem)
On the other hand, this rule may double the original member amount in memo group, things get worse if we join many tables. This overhead of pushing down null rejection filters may cancel out its benefits.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yongyanglai)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n1xpyCfhuUFb9bMJv:-M9n1xpyCfhuUFb9bMJw:b-v8oji) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L88)):*
> ```Go
> 		if expr, ok := in.(*memo.InnerJoinExpr); ok {
> 			notNullColsFromFilters := ExtractAllNotNullColsFromFilters(evalCtx, expr.On)
> 			relProps.Rule.RejectNullCols.Difference(notNullColsFromFilters)
> ```

This line is a no-op.. did you mean DifferenceWith? Can you add a comment explaining why we're doing this? Is it that the join itself would not benefit from not-null filters on the equality columns because it requires not-null values on them anyway?  If that's what we're doing, it would be better to just DifferenceWith the NotNullCols that were already calculated for the operator (and which include the equality columns for inner joins). And same for the Select case; in fact, we may want to do it for all operators (after the `switch`) just in case.
___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n2BMECwo98Iaw0xOR:-M9n2BMECwo98Iaw0xOS:bcbvjv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L91)):*
> ```Go
> 		}
> 		if expr, ok := in.(*memo.InnerJoinApplyExpr); ok {
> 			notNullColsFromFilters := ExtractAllNotNullColsFromFilters(evalCtx, expr.On)
> ```

These two cases can be the same, just use `in.Child(2).(*memo.FiltersExpr)`


<!-- Sent from Reviewable.io -->

yongyanglai commented:
>
> _[pkg/sql/opt/norm/reject_nulls_funcs.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n1xpyCfhuUFb9bMJv:-M9n1xpyCfhuUFb9bMJw:b-v8oji) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L88)):_
> 
> > ```go
> > 		if expr, ok := in.(*memo.InnerJoinExpr); ok {
> > 			notNullColsFromFilters := ExtractAllNotNullColsFromFilters(evalCtx, expr.On)
> > 			relProps.Rule.RejectNullCols.Difference(notNullColsFromFilters)
> > ```
> 
> This line is a no-op.. did you mean DifferenceWith? Can you add a comment explaining why we're doing this? Is it that the join itself would not benefit from not-null filters on the equality columns because it requires not-null values on them anyway? If that's what we're doing, it would be better to just DifferenceWith the NotNullCols that were already calculated for the operator (and which include the equality columns for inner joins). And same for the Select case; in fact, we may want to do it for all operators (after the `switch`) just in case.
> 
@RaduBerinde 
Oh, it's a mistaken, I did mean DifferenceWith, I should have added some tests for relational properties.
I am not sure I get you, what do you mean by "it requires not-null values on them anyway"
Let me explain what I am thinking more detailly, there is the comments on RejectNullCols:

> RejectNullCols is the subset of nullable output columns that can
> potentially be made not-null by one of the RejectNull normalization
> rules.

We can see RejectNullCols is used by upper expression to derive null rejection filters. Any  filter like `x > 1`, `x=y`, `x is not null` in inner join(`ExtractAllNotNullColsFromFilters `considers all these filters) ensure output columns  referred have no null values(this is not the case in outer join). I think upper filters won't benefits from pushing down null rejection filters for these columns and in case of some rule. Besides some normalization rule makes such change(though result may rewritten by other rules immediately, but I can't argue that):
[1]`SELECT * FROM (t1 INNER JOIN t2 ON t1.x = t2.x) WHERE t1.y IS NOT NULL`
=>
[2]`t1 INNER JOIN t2 ON t1.x=t2.x AND t1.y IS NOT NULL`
If  filter in [1] is null rejection pushed down by some rule ([PushDownNullRejectionForJoin]  for example) and it is transformed into [2] and `t1.y IS NOT NULL` is not pushed down by any other rule, [PushDownNullRejectionForJoin] may lead to an indefinite loop and exhaust the memory.
I know the situation I described won't happen for now since [2] will be rewritten by existing rules. But it will be safer to narrow down RejectNullCols in inner join.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yongyanglai)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n1xpyCfhuUFb9bMJv:-M9ncm2zEYiq4d9Vxoz4:bp4sdox) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L88)):*
<details><summary><i>Previously, yongyanglai (YONGYANG LAI) wrote…</i></summary><blockquote>

> > ```go
> > 		if expr, ok := in.(*memo.InnerJoinExpr); ok {
> > 			notNullColsFromFilters := ExtractAllNotNullColsFromFilters(evalCtx, expr.On)
> > 			relProps.Rule.RejectNullCols.Difference(notNullColsFromFilters)
> > ``` 

>  This line is a no-op.. did you mean DifferenceWith? Can you add a comment explaining why we're doing this? Is it that the join itself would not benefit from not-null filters on the equality columns because it requires not-null values on them anyway? If that's what we're doing, it would be better to just DifferenceWith the NotNullCols that were already calculated for the operator (and which include the equality columns for inner joins). And same for the Select case; in fact, we may want to do it for all operators (after the `switch`) just in case.
> 
> 

@RaduBerinde  Oh, it's a mistaken, I did mean DifferenceWith, I should have added some tests for relational properties. I am not sure I get you, what do you mean by "it requires not-null values on them anyway" Let me explain what I am thinking more detailly, there is the comments on RejectNullCols: 

> RejectNullCols is the subset of nullable output columns that can potentially be made not-null by one of the RejectNull normalization rules. 

We can see RejectNullCols is used by upper expression to derive null rejection filters. Any  filter like `x > 1`, `x=y`, `x is not null` in inner join(`ExtractAllNotNullColsFromFilters `considers all these filters) ensure output columns  referred have no null values(this is not the case in outer join). I think upper filters won't benefits from pushing down null rejection filters for these columns . Besides some normalization rule makes such change(though result may rewritten by other rules immediately, but I can't argue that): [1]`SELECT * FROM (t1 INNER JOIN t2 ON t1.x = t2.x) WHERE t1.y IS NOT NULL` => [2]`t1 INNER JOIN t2 ON t1.x=t2.x AND t1.y IS NOT NULL` If  filter in [1] is null rejection pushed down by some rule ([PushDownNullRejectionForJoin]  for example) and it is transformed into [2] and `t1.y IS NOT NULL` is not pushed down by any other rule, [PushDownNullRejectionForJoin] may lead to an indefinite loop and exhaust the memory because original RejectNullCols in inner join didn't reflect these filters. I know the situation I described won't happen for now since [2] will be rewritten by existing rules. But it will be safer to narrow down RejectNullCols in inner join. 
</blockquote></details>

Yeah. In a more general way, we can say that output columns which are already known to be not null (like `x,y` in your example) should never be RejectNullCols. We can just apply that logic (`relProps.Rule.RejectNullCols.DifferenceWith(relProps.NotNullCols)`) to all operators and it will automatically do what you are doing (we are already looking at the filters when deriving NotNullCols for InnerJoin and Select).

In any case, with this fix (Difference -> DifferenceWith), do you think it's worth checking again if making this a norm rule would work? I am guessing this bug causes many unnecessary filters to show up.


<!-- Sent from Reviewable.io -->

yongyanglai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yongyanglai)
> 
> _[pkg/sql/opt/norm/reject_nulls_funcs.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n1xpyCfhuUFb9bMJv:-M9ncm2zEYiq4d9Vxoz4:bp4sdox) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L88)):_
> 
> Previously, yongyanglai (YONGYANG LAI) wrote…
> Yeah. In a more general way, we can say that output columns which are already known to be not null (like `x,y` in your example) should never be RejectNullCols. We can just apply that logic (`relProps.Rule.RejectNullCols.DifferenceWith(relProps.NotNullCols)`) to all operators and it will automatically do what you are doing (we are already looking at the filters when deriving NotNullCols for InnerJoin and Select).
> 
> In any case, with this fix (Difference -> DifferenceWith), do you think it's worth checking again if making this a norm rule would work? I am guessing this bug causes many unnecessary filters to show up.

I agree using NotNullCols will be more convenient, but I am not sure if it is proper to rely on another relational properties when deriving RejectNullCols and you said it is a best-effort property.
If you think it's ok, we can do that.
I will check again to see if it is possible to make this rule a normalization one. But I doubt it because many previous cases l use to verify the rule are not affected by this nit. Anyway I will give it a try and update code later.

RaduBerinde commented:
___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-M9n1xpyCfhuUFb9bMJv:-M9paHIp-loNOAH5YH7b:brbs9ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/379986bdc163675c4faaaf0893b0d16484eca4f0/pkg/sql/opt/norm/reject_nulls_funcs.go#L88)):*
<details><summary><i>Previously, yongyanglai (YONGYANG LAI) wrote…</i></summary><blockquote>

I agree using NotNullCols will be more convenient, but I am not sure if it is proper to rely on another relational properties when deriving RejectNullCols and you said it is a best-effort property. If you think it's ok, we can do that. I will check again to see if it is possible to make this rule a normalization one. But I doubt it because many previous cases l use to verify the rule are not affected by this nit. Anyway I will give it a try and update code later.   
</blockquote></details>

It is proper to rely on other properties. The code you are adding here is completely redundant with the code that calculates the not null cols for join and select.


<!-- Sent from Reviewable.io -->

yongyanglai commented:
I am still unable to make it a norm rule, pushing down null rejection filters arbitrarily still lead to higher cost in many cases. 
I have revised the code according to previous review and add several tests for property RejectNullCols 
PTAL

andy-kimball commented:
I think we must find a way to make this a normalization rule. Exploration rules are too expensive in terms of planning time to use for something like this. Maybe we could use the NULL count estimation in the statistics to decide whether to push down a NULL filter. Can you show a couple examples of queries that aren't working well with this as a normalization rule?

yongyanglai commented:
> I think we must find a way to make this a normalization rule. Exploration rules are too expensive in terms of planning time to use for something like this. Maybe we could use the NULL count estimation in the statistics to decide whether to push down a NULL filter. Can you show a couple examples of queries that aren't working well with this as a normalization rule?

I agree that exploration rules are too expensive, rendering pushing down null rejection less valuable.
The simplest situation this rule not working well as normalization one is that inputs on both side have little null value. For example, we get a query:
```sql
SELECT * FROM t1 JOIN t2 ON t1.a = t2.x
```
Without pushing down null rejection:
```
inner-join (hash)
 ├── columns: a:1(int!null) b:2(int) c:3(int) x:5(int!null) y:6(int) c:7(int)
 ├── stats: [rows=399960.001, distinct(1)=9999, null(1)=0, distinct(5)=9999, null(5)=0]
 ├── cost: 437999.65
 ├── fd: (1)==(5), (5)==(1)
 ├── prune: (2,3,6,7)
 ├── reject-nulls: (2,3,6,7)
 ├── scan t.public.t1
 │    ├── columns: t.public.t1.a:1(int) t.public.t1.b:2(int) t.public.t1.c:3(int)
 │    ├── stats: [rows=200000, distinct(1)=100000, null(1)=10]
 │    ├── cost: 214000.02
 │    ├── prune: (1-3)
 │    ├── reject-nulls: (1-3)
 │    └── unfiltered-cols: (1-4)
 ├── scan t.public.t2
 │    ├── columns: t.public.t2.x:5(int) t.public.t2.y:6(int) t.public.t2.c:7(int)
 │    ├── stats: [rows=200000, distinct(5)=10000, null(5)=10]
 │    ├── cost: 214000.02
 │    ├── prune: (5-7)
 │    ├── reject-nulls: (5-7)
 │    └── unfiltered-cols: (5-8)
 └── filters
      └── eq [type=bool, outer=(1,5), constraints=(/1: (/NULL - ]; /5: (/NULL - ]), fd=(1)==(5), (5)==(1)]
           ├── variable: t.public.t1.a:1 [type=int]
           └── variable: t.public.t2.x:5 [type=int]
```  
If we push down null rejection:
```
inner-join (hash)
 ├── columns: a:1(int!null) b:2(int) c:3(int) x:5(int!null) y:6(int) c:7(int)
 ├── stats: [rows=399960.001, distinct(1)=100000, null(1)=0, distinct(5)=100000, null(5)=0]
 ├── cost: 441999.37
 ├── fd: (1)==(5), (5)==(1)
 ├── prune: (2,3,6,7)
 ├── reject-nulls: (2,3,6,7)
 ├── select
 │    ├── columns: t.public.t1.a:1(int!null) t.public.t1.b:2(int) t.public.t1.c:3(int)
 │    ├── stats: [rows=199990, distinct(1)=100000, null(1)=0]
 │    ├── cost: 216000.03
 │    ├── prune: (2,3)
 │    ├── reject-nulls: (2,3)
 │    ├── scan t.public.t1
 │    │    ├── columns: t.public.t1.a:1(int) t.public.t1.b:2(int) t.public.t1.c:3(int)
 │    │    ├── stats: [rows=200000, distinct(1)=100000, null(1)=10]
 │    │    ├── cost: 214000.02
 │    │    ├── prune: (1-3)
 │    │    └── reject-nulls: (1-3)
 │    └── filters
 │         └── is-not [type=bool, outer=(1), constraints=(/1: (/NULL - ]; tight)]
 │              ├── variable: t.public.t1.a:1 [type=int]
 │              └── null
 ├── select
 │    ├── columns: t.public.t2.x:5(int!null) t.public.t2.y:6(int) t.public.t2.c:7(int)
 │    ├── stats: [rows=199990, distinct(5)=100000, null(5)=0]
 │    ├── cost: 216000.03
 │    ├── prune: (6,7)
 │    ├── reject-nulls: (6,7)
 │    ├── scan t.public.t2
 │    │    ├── columns: t.public.t2.x:5(int) t.public.t2.y:6(int) t.public.t2.c:7(int)
 │    │    ├── stats: [rows=200000, distinct(5)=100000, null(5)=10]
 │    │    ├── cost: 214000.02
 │    │    ├── prune: (5-7)
 │    │    └── reject-nulls: (5-7)
 │    └── filters
 │         └── is-not [type=bool, outer=(5), constraints=(/5: (/NULL - ]; tight)]
 │              ├── variable: t.public.t2.x:5 [type=int]
 │              └── null
 └── filters
      └── eq [type=bool, outer=(1,5), constraints=(/1: (/NULL - ]; /5: (/NULL - ]), fd=(1)==(5), (5)==(1)]
           ├── variable: t.public.t1.a:1 [type=int]
           └── variable: t.public.t2.x:5 [type=int]
```
You just mind me that we can check statistics properties in normalization rule. But we still need some hints  like per-low cost in join, or we don't know the threshold for pushing down. Maybe we can use some fix coefficient for estimation or maybe we just don't need to worry too much about increased cost(it seems a very little portion of whole cost)

RaduBerinde commented:
That plan change looks good to me.. The estimated cost is slightly higher but I bet it's usually faster in practice, even with low ratio of NULLs. It saves each such row from being hashed.  If we want to check the fraction of NULLs in the stats, the treshold doesn't need to be that accurate (although we could run some experiments).

I was worried there were bigger problems; like the new filters interfering with the application of an important rule.

yongyanglai commented:
I didn't see huge changes in outputs of existing logic test except for properties RejectNullCols and null rejection filters pushed down although many tests use relatively simple statement and they were unable to cover all search paths. I haven't figure out how to verify this, any idea on this?
I will make it a norm rule with stats checking to see if there is any problem

yongyanglai commented:
I change the rule to a normalization rule [AddNullRejectionForJoin]. This rule doesn't push down null rejection filters directly, instead it populate on condition with null rejection filters and wait for other rules to push down. It also make rough estimation of threshold of pushing down.
PTAL

RaduBerinde reviewed (commented):
Thanks for your patience working through this! I like this version overall!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-:-MAK2I5-Cjf3JzSxWCvO:buw27su)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @rytaft, and @yongyanglai)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1GQv93OCYyVJhBR7:-MAK1GQv93OCYyVJhBR8:bpwp07r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/reject_nulls_funcs.go#L193)):*
> ```Go
> // null rejection filters that will not change the output result and will be later pushed
> // down by other rules
> func (c *CustomFuncs) IsAddingNullRejectionFiltersPossible(
> ```

CanAddNullRejectionFilterToJoin or ShouldAddNullRejectionFilterToJoin
___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1cQTD6MM_LQw4RGB:-MAK1cQTD6MM_LQw4RGC:bjfgnn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/reject_nulls_funcs.go#L224)):*
> ```Go
> 			return false
> 		}
> 		// for detail of coefficient, see computeHashJoinCost in coster.go
> ```

I think we should keep things simpler and just define a fraction (e.g. 20%), and only push down filters if there are at least that many nulls.
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0VTEB0iXTLfAX9-M:-MAK0VTEB0iXTLfAX9-N:b-nouil6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L153)):*
> ```
> )
> 
> # AddNullRejectionForJoin derive not-null cols based on current filters in
> ```

[nit] derives
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0XJmDTu0_v53SUw_:-MAK0XJmDTu0_v53SUwa:b-w8n524) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L154)):*
> ```
> 
> # AddNullRejectionForJoin derive not-null cols based on current filters in
> # join operator and populate on condition with null rejection filters. It handles
> ```

the join operator
populates the ON condition
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0ZPb5GR43vbLW5QG:-MAK0ZPb5GR43vbLW5QH:bbvupsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L155)):*
> ```
> # AddNullRejectionForJoin derive not-null cols based on current filters in
> # join operator and populate on condition with null rejection filters. It handles
> # statement like these:
> ```

For example, it transforms this statement:
..
into:
..
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0ib0BVZXhuvkpS0j:-MAK0ib0BVZXhuvkpS0k:b-autety) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L163)):*
> ```
> #   SELECT * FROM t1 JOIN t2 ON t1.x = t2.x AND t1.x IS NOT NULL AND t2.x IS NOT NULL
> #
> # The condition t1.x = t.x implies that t1.x IS NOT NULL AND t.x IS NOT NULL, we
> ```

; we construct these filters and add them to the ON filters. These filters will then be pushed down by other rules.
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 168 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0wCY-rIhelb4KuHJ:-MAK0wCY-rIhelb4KuHK:b-b9kt1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L168)):*
> ```
> # benefits if input contain a lot of null values.
> #
> # This rule applies to three join operator: InnerJoin, LeftJoin, RightJoin. For InnerJoin
> ```

RightJoin does not exist during normalization.
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 169 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK13NV2sKlDC8ofWFi:-MAK13NV2sKlDC8ofWFj:bl02bpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L169)):*
> ```
> #
> # This rule applies to three join operator: InnerJoin, LeftJoin, RightJoin. For InnerJoin
> # there is possibility for us to push down is-not-null filters to both left input and right
> ```

For InnerJoin we push down ..
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK15qf-_SbT26tUCuL:-MAK15qf-_SbT26tUCuM:b-mu9ko) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L170)):*
> ```
> # This rule applies to three join operator: InnerJoin, LeftJoin, RightJoin. For InnerJoin
> # there is possibility for us to push down is-not-null filters to both left input and right
> # input or only singe side. For outer join, we can only push is-not-null filters to another side,
> ```

For left join, we pus is-not-null filters to the left side.
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1A5-5VylGV9lQAWX:-MAK1A5-5VylGV9lQAWY:bedd2kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L182)):*
> ```
> # we lost some output rows whose t1.x is null, while statement [1] remains them. So we don't push down
> # is-not-null filters to another side in outer join
> #
> ```

This text needs to mention that the rule only gets applied if there is a certain fraction of NULLs in the relevant inputs.
___
*[pkg/sql/opt/xform/testdata/external/activerecord, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK24GG8_ylJ5Y0QycV:-MAK24GG8_ylJ5Y0QycW:bxn1com) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cbcde31b0109128291c2fc01f94e6884d0d4052/pkg/sql/opt/xform/testdata/external/activerecord#L131)):*
> ```
>       │    ├── fd: ()-->(1,15), (6)-->(2,3,9,13,18,23-25,27,28,34,61), (2)-->(3,6,9,13,18), (24)-->(25), (27)-->(28), (34)-->(61)
>       │    ├── inner-join (cross)
>       │    │    ├── columns: adrelid:23 adnum:24 adbin:25 c.oid:27!null c.collname:28!null t.oid:34!null typcollation:61!null
> ```

Any idea what explains this change in columns? (and the similar ones in other files)


<!-- Sent from Reviewable.io -->

yongyanglai committed:
opt: populate on condition in join with null rejection filters

Previvously, the optimizer didn't push down null rejection filters
implies by on-condition in join expression. For example:

SELECT * t1 JOIN t2 ON t1.x=t2.x

This expression can be rewritten into:

SELECT * FROM (SELECT * FROM t1 WHERE t1.x IS NOT NULL) AS t1_
JOIN (SELECT * FROM t2 WHERE t2.x IS NOT NULL) AS t2_ ON t1_.x=t2_.x

If columns x in t1 and t2 contains many null values, we may benefits
from it.To acheive this goal, I add a new rewriting rule
[AddNullRejectionForJoin]. The whole work contains three parts:

  1. Add support for operators <Scan> and <Select> to derive null
  rejection columns
  2. Add rule [AddNullRejectionForJoin] and corresponding logical
  test
  3. Update other logical tests

This commit contains first two parts.

Release note: None

yongyanglai committed:
opt: populate on condition in join with null rejection filters

Previvously, the optimizer didn't push down null rejection filters
implies by on-condition in join expression. For example:

SELECT * t1 JOIN t2 ON t1.x=t2.x

This expression can be rewritten into:

SELECT * FROM (SELECT * FROM t1 WHERE t1.x IS NOT NULL) AS t1_
JOIN (SELECT * FROM t2 WHERE t2.x IS NOT NULL) AS t2_ ON t1_.x=t2_.x

If columns x in t1 and t2 contains many null values, we may benefits
from it.To acheive this goal, I add a new rewriting rule
[AddNullRejectionForJoin]. The whole work contains three parts:

  1. Add support for operators <Scan> and <Select> to derive null
  rejection columns
  2. Add rule [AddNullRejectionForJoin] and corresponding logical
  test
  3. Update other logical tests

This commit contains third part.

Release note: None

yongyanglai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, @rytaft, and @yongyanglai)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1GQv93OCYyVJhBR7:-MALocx32k4act5JZbgQ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/reject_nulls_funcs.go#L193)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

CanAddNullRejectionFilterToJoin or ShouldAddNullRejectionFilterToJoin
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1cQTD6MM_LQw4RGB:-MALomSf0_K-WOWg3e-a:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/reject_nulls_funcs.go#L224)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think we should keep things simpler and just define a fraction (e.g. 20%), and only push down filters if there are at least that many nulls.
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0VTEB0iXTLfAX9-M:-MALorA9621g2hLwZGZf:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L153)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] derives
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0XJmDTu0_v53SUw_:-MALzcqe-kQ3VnF-WdYU:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L154)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

the join operator
populates the ON condition
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0ZPb5GR43vbLW5QG:-MALzm-t81Rn6hQLLf3h:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L155)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

For example, it transforms this statement:
..
into:
..
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0ib0BVZXhuvkpS0j:-MALzqyu02PdpuCuSGbM:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L163)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

; we construct these filters and add them to the ON filters. These filters will then be pushed down by other rules.
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 168 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK0wCY-rIhelb4KuHJ:-MAM-Bwd4mcVtVOTuFbL:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L168)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

RightJoin does not exist during normalization.
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 169 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK13NV2sKlDC8ofWFi:-MAM-ISBD8peQv9mnje3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L169)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

For InnerJoin we push down ..
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK15qf-_SbT26tUCuL:-MAM8MyvD4NOqBKnasRU:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L170)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

For left join, we pus is-not-null filters to the left side.
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/reject_nulls.opt, line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK1A5-5VylGV9lQAWX:-MAM8Ok65zDTzufMCRIi:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf6706584ba850e4f6d65eba26903556b540158/pkg/sql/opt/norm/rules/reject_nulls.opt#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This text needs to mention that the rule only gets applied if there is a certain fraction of NULLs in the relevant inputs.
</blockquote></details>

Done
___
*[pkg/sql/opt/xform/testdata/external/activerecord, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK24GG8_ylJ5Y0QycV:-MAM8PwM3s0SfktEFD9U:b-ymisbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cbcde31b0109128291c2fc01f94e6884d0d4052/pkg/sql/opt/xform/testdata/external/activerecord#L131)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Any idea what explains this change in columns? (and the similar ones in other files)
</blockquote></details>

Thanks for reviews!
I think change in "activerecord" is a mistake, but changes  in "navicat" and "liquibase" are not.
Change in "activerecord" is caused by directly assigning `opt.ColSet` in original code of `DeriveRejectNullCols`:
```go
case opt.RightJoinOp:
// Pass through null-rejection columns from right input, and request null-
// rejection on left columns.
relProps.Rule.RejectNullCols = in.Child(0).(memo.RelExpr).Relational().OutputCols
if in.Child(1).(memo.RelExpr).Relational().OuterCols.Empty() {
relProps.Rule.RejectNullCols.UnionWith(DeriveRejectNullCols(in.Child(1).(memo.RelExpr)))
}
```
If `ColSet` contains column ID greater than 63, assigning it directly only copy the underlying pointer, thus `RejectNullCols` and `OutputCols` refer to same column set. Computations happened here also change `OutputCols` of child expression. This caused infinite loop when I was testing. So I replaced other direct assignments with `UnionWith ` in `DeriveRejectNullCols` but I omited this line, which cause changes in "activerecord".

Changes in  "navicat" and "liquibase" are caused by the same problem. `logicalPropsBuilder` assign child's `OutputCols` to father's `OutputCols` directly and then father's `OutputCols` is assigned to `RejectNullCols` directly. If column set contains column ID greater than 63, then computation on `RejectNullCols` cause all these expression's `OutputCols` change. So I think previous output columns in "navicat" and "liquibase" are incorrect, they contain sibling expression's columns. I have checked these disappeared output columns and I think they conform to changes in  `DeriveRejectNullCols`


<!-- Sent from Reviewable.io -->
PTAL

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @rytaft, and @yongyanglai)

___
*[pkg/sql/opt/xform/testdata/external/activerecord, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50186#-MAK24GG8_ylJ5Y0QycV:-MAMYtst817CkZqawjDj:buk34pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cbcde31b0109128291c2fc01f94e6884d0d4052/pkg/sql/opt/xform/testdata/external/activerecord#L131)):*
<details><summary><i>Previously, yongyanglai (YONGYANG LAI) wrote…</i></summary><blockquote>

Thanks for the review!
I think change in "activerecord" is a mistake, but changes  in "navicat" and "liquibase" are not.
Change in "activerecord" is caused by directly assigning `opt.ColSet` in original code of `DeriveRejectNullCols`:
```go
case opt.RightJoinOp:
// Pass through null-rejection columns from right input, and request null-
// rejection on left columns.
relProps.Rule.RejectNullCols = in.Child(0).(memo.RelExpr).Relational().OutputCols
if in.Child(1).(memo.RelExpr).Relational().OuterCols.Empty() {
relProps.Rule.RejectNullCols.UnionWith(DeriveRejectNullCols(in.Child(1).(memo.RelExpr)))
}
```
If `ColSet` contains column ID greater than 63, assigning it directly only copy the underlying pointer, thus `RejectNullCols` and `OutputCols` refer to same column set. Computations happened here also change `OutputCols` of child expression. This caused infinite loop when I was testing. So I replaced other direct assignments with `UnionWith ` in `DeriveRejectNullCols` but I omited this line, which cause changes in "activerecord".

Changes in  "navicat" and "liquibase" are caused by the same problem. `logicalPropsBuilder` assign child's `OutputCols` to father's `OutputCols` directly and then father's `OutputCols` is assigned to `RejectNullCols` directly. If column set contains column ID greater than 63, then computation on `RejectNullCols` cause all these expression's `OutputCols` change. So I think previous output columns in "navicat" and "liquibase" are incorrect, they contain sibling expression's columns. I have checked these disappeared output columns and I think they conform to changes in  `DeriveRejectNullCols`
</blockquote></details>

Oh, wow, that's a bad bug. Nice find! We should fix that separately (and backport it).

Interestingly, we run tests with `smallCutoff` in `util/fast_int_set.go` set to something very small for each release to find similar bugs; but here we would have needed the opposite (an implementation that always behaves like in the "small" case).

I can make a separate PR for this fix, unless you prefer to do it? I reproduced test differences with this fix:
```
+++ b/pkg/sql/opt/norm/reject_nulls_funcs.go
@@ -93,10 +93,10 @@ func DeriveRejectNullCols(in memo.RelExpr) opt.ColSet {
 	case opt.RightJoinOp:
 		// Pass through null-rejection columns from right input, and request null-
 		// rejection on left columns.
-		relProps.Rule.RejectNullCols = in.Child(0).(memo.RelExpr).Relational().OutputCols
 		if in.Child(1).(memo.RelExpr).Relational().OuterCols.Empty() {
 			relProps.Rule.RejectNullCols.UnionWith(DeriveRejectNullCols(in.Child(1).(memo.RelExpr)))
 		}
+		relProps.Rule.RejectNullCols.UnionWith(in.Child(1).(memo.RelExpr).Relational().OutputCols)
```


<!-- Sent from Reviewable.io -->

yongyanglai commented:
> Interestingly, we run tests with `smallCutoff` in `util/fast_int_set.go` set to something very small for each release to find similar bugs; but here we would have needed the opposite (an implementation that always behaves like in the "small" case).
> 
> I can make a separate PR for this fix, unless you prefer to do it? I reproduced test differences with this fix:
> 
I'd like to do that, but how do you want it be fixed? Do you want only misusages in `DeriveRejectNullCols` to be fixed or you want find all misusages in project and fix them. Besides, to which version should we backport the fix?

RaduBerinde commented:
Do you know of any other misuses outside of DeriveRejectNullCols? It's ok for expressions to share `OutputCols` because they are not supposed to be modified once put in the properties.

RaduBerinde commented:
As for the backport, we should backport to 20.1, 19.2, 19.1. We use the https://github.com/benesch/backport utility for easy backporting. I can also take care of the backports.

rytaft commented:
https://github.com/cockroachdb/cockroach/pull/51966 was recently merged to add null-rejection filters under joins. Should this PR be closed now?





<!-- Sent from Reviewable.io -->

yongyanglai commented:
sorry for keeping this inactive PR opened for so long, I will close it now.
